shared/src/main/scala/effekt/symbols/Name.scala

```diff
@@ -3,27 +3,133 @@ package effekt.symbols
+ * 
+ * @note Don't use the constructor of this type directly. Instead use [[Name(qualifiedName)]] to safely convert strings to names.
+ */
+case class ToplevelName(localName: String) extends Name {
+  def parentOption: Option[Name] = None
+
+  def qualifiedName: String = localName
+
+  def rename(f: String => String): Name = ToplevelName(f(localName))
 }
 
-case object NoName extends Name {
-  def name = "<NoName>"
-  def qualified: String = name
-  def rename(f: String => String): Name = this
+/**
+ * A Name which is a child of a other name, e.g. the name of a nested module.
+ *
+ * @param parent The parent of this name.
+ * @param localName The local name relative to the parent.
+ * The local name should not contain point characters (`'.'`). This could lead to unexpected behavior.
+ * 
+ * @note Creation of [[NestedName]] via the [[Name.nested()]] method is prefered.
+ */
+case class NestedName(parent: Name, localName: String) extends Name {
+  def parentOption: Some[Name] = Some(parent)
+
+  def qualifiedName: String = s"${parent.qualifiedName}.$localName"
+
+  def rename(f: String => String): Name = NestedName(parent, f(localName))
 }
 
+// Pseudo-constructors to safely convert ids and strings into names.
 object Name {
-  def apply(id: Id)(implicit C: Context): Name = QualifiedName(id.name, C.module)
-  def apply(name: String, module: Module) = QualifiedName(name, module)
+  /**
+   * Creates a name for the ID inside the module of the current context.
+   *
+   * @param id The id to be converted into a name.
+   * @param C The implicit context used to find the current module.
+   * @return A qualified name inside the module.
+   */
+  def apply(id: Id)(implicit C: Context): Name = Name.apply(id.name, C.module)
+
+  /**
+   * Creates a qualified name for a symbol inside the given module.
+   *
+   * @param name the local name of the symbol.
+   * @param module the module containing the symbol.
+   * @return A name with the module path as a parent.
+   *
+   * @example The name `"baz"` inside the module with the path `"foo/bar"` is parsed as qualified name `"foo.bar.baz"`
+   */
+  def apply(name: String, module: Module): Name = module.name.nested(name)
+
+  /**
+   * Parses a [[Name]] from a String which might be a qualified name.
+   * @param name A non-empty String which might be a qualified or local name.
+   * @return A Name where [[Name.qualifiedName]] is equal to the input name.
+   *
+   * @example `Name("foo") == ToplevelName("foo")`
+   * @example `Name("foo.bar.baz") == NestedName(NestedName(ToplevelName("foo"), "bar"), "baz")`
+   */
+  def apply(name: String): Name = {
+    assert(name.nonEmpty, "Name cannot be empty.")
+
+    // Is nested?
+    if (name.contains('.')) {
+      val segments = name.split('.')
+      val top: Name = ToplevelName(segments.head)
+      segments.drop(1).foldLeft(top)((parent, segment) => parent.nested(segment))
+    } else {
+      ToplevelName(name)
+    }
```

Comment:
Why do you have to distinguish the two cases? I would assume that something like:

```
  val (top :: rest) = name.split('.').toList
  rest.foldLeft(ToplevelName(top)) { _ nested _ }
```
would do the job.